File: solutions/elixir/rotational-cipher/1/lib/rotational_cipher.ex
defmodule RotationalCipher do
  @doc """
  Given a plaintext and amount to shift by, return a rotated string.

  Example:
  iex> RotationalCipher.rotate("Attack at dawn", 13)
  "Nggnpx ng qnja"
  """
  @end_of_english_alphabet 26
  @start_of_ascii_uppercase_alphabet 65
  @end_of_ascii_uppercase_alphabet 90
  @start_of_ascii_lowercase_alphabet 97
  @end_of_ascii_lowercase_alphabet 122

  @spec rotate(text :: String.t(), shift :: integer) :: String.t()
  def rotate(text, shift),
    do:
      text
      |> to_charlist()
      |> Enum.map(&shift_character(&1, shift))
      |> to_string()

  def shift_character(character, shift)
      when shift >= @end_of_english_alphabet or
             (character <= @end_of_ascii_uppercase_alphabet and
                character + shift > @end_of_ascii_uppercase_alphabet) or
             character + shift > @end_of_ascii_lowercase_alphabet,
      do: character + shift - @end_of_english_alphabet

  def shift_character(character, _shift)
      when character < @start_of_ascii_uppercase_alphabet or
             (character > @end_of_ascii_uppercase_alphabet and
                character < @start_of_ascii_lowercase_alphabet) or
             character > @end_of_ascii_lowercase_alphabet,
      do: character

  def shift_character(character, shift),
    do: character + shift
end

File: solutions/scala/rotational-cipher/1/src/main/scala/RotationalCipher.scala
object RotationalCipher:
    def rotate(text: String, shift: Int): String =
        val normalizedShift = shift % 26
        val rotateCharacter  = applyRotation(normalizedShift)
        text.map(rotateCharacter )

    private def applyRotation(shift: Int)(character: Char): Char =
        if character >= 'a' && character <= 'z'
        then shiftToLowercase(shift, character)
        else if character >= 'A' && character <= 'Z'
             then shiftToUppercase(shift, character)
             else character

    private def shiftToUppercase(shift: Int, character: Char): Char =
        rotateWithinAlphabet(shift, character, 'A')

    private def shiftToLowercase(shift: Int, character: Char): Char =
        rotateWithinAlphabet(shift, character, 'a')

    private def rotateWithinAlphabet(shift: Int, character: Char, alphabetStart: Char): Char =
        val base = alphabetStart.toInt
        val offset = character.toInt - base + shift
        (offset % 26 + base).toChar